### ISSUE TITLE
python - einsum scalar return type

### ISSUE THREAD

**Issue by jcmgray** [cyclops-community/ctf#105]
Sometimes in automatic tensor network contractions equations like
```python
import ctf

x = ctf.random.random((2, 2, 2))
y = ctf.random.random((2, 2, 2))
z = ctf.random.random((2,))

xy = ctf.einsum('abc,abc->', x, y)
# xy = ctf.tensordot(x, y, 3)  # works
# xy = ctf.astensor(ctf.einsum('abc,abc->', x, y))  # works
xyz = ctf.einsum(',a->a', xy, z)
# AttributeError as xy is a float not a ctf.tensor
```
are generated, which is to say, a scalar is produced then multiplied into something. This is fine if ``xy`` is a ``ctf.tensor`` with ``shape=()``. However if ``ctf.einsum`` was itself used to generate ``xy`` (as above) then it gets converted to a native python type rather than staying a ``ctf.tensor`` and so must be handled manually.

From my perspective it would be cleanest to have ``ctf.einsum`` always return a ``ctf.tensor``, but look's like its an intentional choice not to in the 0D case:

https://github.com/cyclops-community/ctf/blob/c51cd841ecefedbcdb1593b702f3ea6d2ecb380a/src_python/ctf/core.pyx#L5606-L5608

That's also obviously pretty reasonable for many uses, but just wondering if you are wedded to it.

**Comment by solomonik** [comment]
Looked into this and I think our current choice is made to match the numpy convention for the einsum output being a scalar number in that case as opposed to a 0 order numpy array storing a scalar. I imagine if we did not, it would complicate conversion of numpy programs to CTF programs.

However, I just pushed a commit that will at least enable CTF's einsum to automatically convert stuff that is passed in to a CTF tensor if its not, which should make the code in your example work. Some other functions have similar handling already while others may need similar fix-up, but I guess its at least consistent with numpy.
